File: 프로그래머스/lv3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    class Node {
        int vertex;
        int cost;
        
        public Node(int vertex, int cost) {
            this.vertex = vertex;
            this.cost = cost;
        }
        
        public String toString() {
            return "vertex = " + this.vertex + ", cost = " + this.cost;
        }
    }
    
    public int solution(int n, int[][] edge) {
        List<List<Integer>> list = new ArrayList<>();
        for (int i=0; i<=n; i++) {
            list.add(new ArrayList());
        }
        
        for (int i=0; i<edge.length; i++) {
            list.get(edge[i][0]).add(edge[i][1]);
            list.get(edge[i][1]).add(edge[i][0]);
        }
        
        int[] distArr = bfs(list, 1);
        
        int max = 0;
        int answer = 0;
        for (int i=0; i<distArr.length; i++) {
            if (max == distArr[i]) {
                answer++;
                continue;
            }
            
            if (max > distArr[i]) {
                continue;
            }
            
            max = Math.max(distArr[i], max);
            answer = 1;
        }
        
        return answer;
    }
    
    private int[] bfs(List<List<Integer>> list, int start) {
        int[] dist = new int[list.size()];
        Queue<Node> queue = new LinkedList<>();
        queue.add(new Node(start, 0));
        dist[start] = 0;
        
        while (!queue.isEmpty()) {
            Node curr = queue.poll();
            
            for (int i=0; i<list.get(curr.vertex).size(); i++) {
                int next = list.get(curr.vertex).get(i);
                
                if (dist[next] != 0 || next == 1) continue;
                
                queue.add(new Node(next, curr.cost+1));
                dist[next] = curr.cost+1;
            }
        }

        return dist;
    }
}

File: 프로그래머스/lv3/49189. 가장 먼 노드/가장 먼 노드.kt
import java.util.*

class Solution {
    companion object {
        private const val START_NODE = 1
    }
    
    fun solution(n: Int, edge: Array<IntArray>): Int {
        val edges: MutableList<MutableList<Int>> = mutableListOf()
        
        for (i in 0 .. n) {
            edges.add(mutableListOf())
        }
        
        for (i in edge.indices) {
            edges.get(edge[i][0]).add(edge[i][1])
            edges.get(edge[i][1]).add(edge[i][0])
        }
        
        val distanceArray: IntArray = djkstra(edges, START_NODE, n)
        
        var maxDistance: Int = 0
        
        for (i in distanceArray.indices) {
            if (distanceArray[i] == Int.MAX_VALUE) continue
            
            if (distanceArray[i] > maxDistance) {
                maxDistance = distanceArray[i]
            }
        }
        
        val maxDistanceCount = distanceArray.count { it == maxDistance }
        
        return maxDistanceCount
    }
    
    fun djkstra(
        edges: MutableList<MutableList<Int>>,
        startNode: Int,
        nodeCount: Int
    ): IntArray {
        val priorityQueue: PriorityQueue<Node> = PriorityQueue() {
                a, b -> a.distance - b.distance
        }

        val distance: IntArray = IntArray(nodeCount + 1)

        for (i in distance.indices) {
            distance[i] = Int.MAX_VALUE
        }

        priorityQueue.add(Node(startNode, 0))
        distance[startNode] = 0

        while (!priorityQueue.isEmpty()) {
            val currentNode: Node = priorityQueue.poll()

            if (currentNode.distance > distance[currentNode.number]) {
                continue
            }

            val nextNodes = edges.get(currentNode.number)

            for (i in nextNodes.indices) {
                val nextNode = edges.get(currentNode.number)[i]

                if (currentNode.distance + 1 < distance[nextNode]) {
                    priorityQueue.add(Node(nextNode, currentNode.distance + 1))
                    distance[nextNode] = currentNode.distance + 1
                }
            }
        }

        return distance
    }
}

data class Node(
    val number: Int,
    val distance: Int = 1
)